temporal-sdk/src/main/java/io/temporal/internal/worker/TrackingSlotSupplier.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal.worker;
+
+import com.uber.m3.tally.Scope;
+import io.temporal.worker.MetricsType;
+import io.temporal.worker.tuning.*;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Wraps a slot supplier and supplements it with additional tracking information that is useful to
+ * provide to all implementations. This type is used internally rather than {@link SlotSupplier}
+ * directly.
+ *
+ * @param <SI> The slot info type
+ */
+public class TrackingSlotSupplier<SI extends SlotInfo> {
+  private final SlotSupplier<SI> inner;
+  private final AtomicInteger issuedSlots = new AtomicInteger();
+  private final Map<SlotPermit, SI> usedSlots = new ConcurrentHashMap<>();
+  private final Scope metricsScope;
+
+  public TrackingSlotSupplier(SlotSupplier<SI> inner, Scope metricsScope) {
+    this.inner = inner;
+    this.metricsScope = metricsScope;
+    publishSlotsMetric();
+  }
+
+  public SlotPermit reserveSlot(SlotReservationData dat, long timeout, TimeUnit timeUnit)
```

Comment:
Sorry this is my fault I should have clarified I think you need 3 over loads here `reserveSlot ` that takes no timeout, `tryReserveSlot` with not timeout that is non blocking, and `tryReserveSlot ` that takes a timeout, just like https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/Lock.html or other interfaces in Java

Reply:
Yeah, kinda realized that after the fact